### ISSUE TITLE
✅ test(e2e): add GPU E2E loss-curve correctness test against d1 reference

### ISSUE THREAD

**Issue by nishide-dev** [nishide-dev/unturtle-legacy#90]
## 背景

現状の E2E テスト (`tests/test_e2e_integration.py`) は「CPU で forward/backward が通る」レベルに留まっている。
dLLM 学習が **正確に** 機能していることを保証するためには、参照実装 (d1 / MDLM) と
loss curve の数値一致を確認するテストが必要。

特に以下の変更が正しく効いているかが未確認:

- `n_maskable` 正規化（Gotcha #15 — 旧 `n_masked` との差が最大 6.7x）
- `DiffusionTrainer` の compute_loss が d1 SFT と同一の loss 値を返すか
- A2D 双方向 attention が学習時に情報リークしないか（causal vs bidirectional の境界）

## 目的

`DiffusionTrainer` の数 step 学習で loss が参照実装と同等に降下することを GPU テストで保証する。

## Acceptance Criteria

- [ ] `tests/diffusion/test_e2e_loss_curve.py` を新規作成（`@pytest.mark.gpu`）
- [ ] A2D LLaMA (tiny, randinit) + DiffusionTrainer で 5 step 学習
- [ ] d1 reference (`dev/repos/d1/SFT/sft_trainer.py`) 相当の pure-PyTorch 実装と同一データ・同一 seed で比較
- [ ] 各 step の loss 差が `atol=1e-3` 以内であることを assert
- [ ] `mask_rate=0.15` / `0.5` 両方でテスト
- [ ] `@pytest.mark.gpu` で CUDA 必須とし、`pytest -m gpu` で選択的実行可能

## 実装方針

```python
# dev/repos/d1/SFT/sft_trainer.py の compute_loss 相当を直接 Python で実装し比較
def d1_reference_loss(logits, labels, diffusion_mask):
    masked_labels = labels.clone()
    masked_labels[~diffusion_mask] = -100
    per_token = F.cross_entropy(logits.view(-1, V), masked_labels.view(-1),
                                ignore_index=-100, reduction="none")
    n_maskable = (labels != -100).sum().clamp_min(1)
    return per_token.sum() / n_maskable
```

## 関連

- CLAUDE.md Gotcha #15: n_maskable 正規化
- `tests/test_e2e_integration.py` (既存 CPU E2E)
- `tests/diffusion/test_gpu_accuracy.py` (既存 kernel 精度テスト)
- `dev/repos/d1/SFT/sft_trainer.py`
